agent/consultemplate.go

```diff
@@ -0,0 +1,65 @@
+package agent
+
+import (
+	"log"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/pkg/errors"
+
+	consultemplateconfig "github.com/hashicorp/consul-template/config"
+	"github.com/hashicorp/consul-template/manager"
+)
+
+type ConsulTemplateConfig struct {
+	Source      string
+	Destination string
+}
+
+func StartConsulTemplate(a *Agent, c *ConsulTemplateConfig) (*manager.Runner, error) {
+	config := consultemplateconfig.DefaultConfig()
+	*config.Templates = append(
+		*config.Templates,
+		&consultemplateconfig.TemplateConfig{
+			Source:      &c.Source,
+			Destination: &c.Destination,
+		},
+	)
+
+	runner, err := manager.NewRunner(config, false)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not start consul-template")
+	}
+	go runner.Start()
+
+	// Create new thread method to check for consul-template events. If the template is rendered
+	// reload the configuration
+	go func() {
+		signals := make(chan os.Signal)
+		signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+		for {
+			select {
+			case quit := <-signals:
+				log.Printf("Exiting from consul-template event listener loop: %s", quit)
+				break
```

Comment:
I don't think we need to handle signals again here. This was done already by the `start()` command handler!
The `Agent` struct has a `ctx` field that you can use to control children goroutines.
This means that instead of waiting for signals in the first `case` statement here, you would instead wait for `<-a.ctx.Done()`, like it is done in `Agent.startCheckTicker()`

Reply:
I forgot to mention: as I implied with my initial review comment, this entire `for` loop should not be run into a goroutine. The only thing we need a goroutine for is `runner.Start()`. The Consul config reload cycle should instead block the entire execution, and we handle concurrency from the outside with the Context and the WaitGroup already available in `Agent.Start()`.